### PR TITLE
Dev/carson/lane detection

### PR DESCRIPTION
**Pull request by DongJunJin**

### CODE REVIEW

File: rr_computer_vision/include/lane_detection.hpp [hunk 1]
```diff
@@ -6,22 +6,71 @@
...
 
 class LaneDetection
 {
-  private:
-
   public:
-    LaneDetection();
+    LaneDetection(ros::NodeHandle nh);
     ~LaneDetection();
+    void Multithreshold(const cv::Mat &input_image, const cv::Mat &bounds, cv::Mat &output_image);
```
Comment (DongJunJin) [inline rr_computer_vision/include/lane_detection.hpp]: Are these functions ever called outside of this class? If not please make them private functions

File: rr_computer_vision/include/lane_detection.hpp [hunk 2]
```diff
@@ -6,22 +6,71 @@
...
+
+    cv::Mat im_input_;
+    cv::Mat Im1_HSV_;
+    cv::Mat Im1_HSV_warped_;
+    cv::Mat transform_matrix_;
+    cv::Mat multibounds_;
+    cv::Mat mask_warped_1_;
+    cv::Mat mask_warped_2_;
+    cv::Mat mask_;
+    cv::Mat out;
```
Comment (DongJunJin) [inline rr_computer_vision/include/lane_detection.hpp]: All internals variables of a class should be have the '_' after it based on our coding style

File: rr_computer_vision/include/lane_detection.hpp [hunk 3]
```diff
@@ -6,22 +6,71 @@
...
+    cv::Mat src;
```
Comment (DongJunJin) [inline rr_computer_vision/include/lane_detection.hpp]: Some of these variables dont really mean much. I understand that theyre storing values of a matrix but what is this acutal matrix? Provide some context

File: rr_computer_vision/src/lane_detection.cpp [hunk 4]
```diff
@@ -20,4 +33,144 @@ LaneDetection::LaneDetection() {
...
+}
+
+void LaneDetection::RGBCameraCallback(const sensor_msgs::Image& msg){
+
+    // Convert sensor_msgs::Image to a BGR8 cv::Mat
+    cv_bridge::CvImagePtr cv_bridge_bgr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
+    cv::Mat img_bgr8 = cv_bridge_bgr->image;
+    cv::cvtColor(img_bgr8, Im1_HSV_, CV_BGR2HSV, 3);
+
+    //src = (cv::Mat_<float>(4,2) << 500.0, 400.0, 750.0, 400.0, 1280.0, 600.0, 0, 600.0);
```
Comment (DongJunJin) [inline rr_computer_vision/src/lane_detection.cpp]: Remove this if it doesnt do anything or provides no description

File: rr_computer_vision/src/lane_detection.cpp [hunk 5]
```diff
@@ -20,4 +33,144 @@ LaneDetection::LaneDetection() {
...
+void LaneDetection::InitializeSubscribers() {
+    zed_subscriber = nh_.subscribe("/zed/rgb/image_rect_color", 1, &LaneDetection::RGBCameraCallback, this);
+}
+
+void LaneDetection::InitializePublishers() {
+    test_publisher = nh_.advertise<sensor_msgs::Image>("/test_publisher", 1);
+    pointList_pub_ = nh_.advertise<nav_msgs::OccupancyGrid>("/point_vec_out_", 1);
+}
+
+void LaneDetection::RGBCameraCallback(const sensor_msgs::Image& msg){
```
Comment (DongJunJin) [inline rr_computer_vision/src/lane_detection.cpp]: Provide some explaination for this function theres a nice layout you can follow

File: rr_computer_vision/src/lane_detection.cpp [hunk 6]
```diff
@@ -20,4 +33,144 @@ LaneDetection::LaneDetection() {
...
+      {
+        occupancy_.push_back(-1);
+      }
+      else
+      {
+        occupancy_.push_back(100);
+      }
+    }
+    grid_msg_.data = occupancy_;
+    meta_data_.height = 720;
```
Comment (DongJunJin) [inline rr_computer_vision/src/lane_detection.cpp]: Have you confirmed that these parameters will always be constant? Are they impacted by anything?

File: rr_computer_vision/src/lane_detection.cpp [hunk 7]
```diff
@@ -20,4 +33,144 @@ LaneDetection::LaneDetection() {
...
+    meta_data_.resolution = 0.01;
+    meta_data_.origin.position.x = 0;
+    meta_data_.origin.position.y = 0;
+
+    grid_msg_.info = meta_data_;
+    pointList_pub_.publish(grid_msg_);
+}
+
+
+void LaneDetection::Multithreshold(const cv::Mat &input_image, const cv::Mat &bounds, cv::Mat &output_image) {
```
Comment (DongJunJin) [inline rr_computer_vision/src/lane_detection.cpp]: Ditto here for providing some descriptions

File: rr_computer_vision/include/lane_detection.hpp [hunk 8]
```diff
@@ -6,22 +6,71 @@
...
+
+    int adapt_hsv_patch_size_;
+    int blob_size_;
+
+    ros::Subscriber zed_subscriber;
+    ros::Publisher test_publisher;
+    ros::Publisher pointList_pub_;
+    ros::NodeHandle nh_;
+
+    cv::Mat im_input_;
```
Comment (DongJunJin) [inline rr_computer_vision/include/lane_detection.hpp]: Are you sure you need all of these variables to be global within the class?
Is there no nice way of making some of these local variables instead?
As long as you understand what comes in and out of the function this should be a relatively easy modification to do.

File: rr_computer_vision/src/lane_detection.cpp (hunk 5 above)
Comment (DongJunJin) [inline rr_computer_vision/src/lane_detection.cpp]: Also I did not mention this before and I take partial responsibility for it but do not make callbacks this long. Our callbacks should be very short (only copying data) and these other stuff should not be occuring in the callback.

File: rr_computer_vision/src/lane_detection.cpp (hunk 6 above)
Comment (CXYCarson) [reply rr_computer_vision/src/lane_detection.cpp]: I'm not what those parameters for an occupancy grid represents. I copied them from last year, could you give me some explanations on what they do?

**Pull request closed without merging**